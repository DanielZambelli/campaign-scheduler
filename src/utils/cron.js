const CronJob = require('cron').CronJob

const cron = (cronStr, callback, autoStart = true) => {
  return new CronJob(cronStr, callback ,null, autoStart)
}

const cronOnce = (cronStr, callback, autoStart = true) => {
  const job = cron(cronStr, async () => {
    await callback()
    job.stop()
  }, autoStart)
  return job
}

module.exports = {cron,cronOnce}
