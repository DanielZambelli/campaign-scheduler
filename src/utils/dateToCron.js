const dateToCron = (dateString) => {
  const date = new Date(dateString)
  const seconds = date.getSeconds()
  const minutes = date.getMinutes()
  const hours = date.getHours()
  const days = date.getDate()
  const months = date.getMonth()
  const dayOfWeek = date.getDay()
  return `${seconds} ${minutes} ${hours} ${days} ${months} ${dayOfWeek}`
}

module.exports = {dateToCron}
