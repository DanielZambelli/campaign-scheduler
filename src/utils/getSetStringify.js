const getSetStringify = (field) => {
  return {
    get(){
      const value = this.getDataValue(field)
      return value ? JSON.parse(value) : value
    },
    set(value){
      if(value) value = JSON.stringify(value)
      this.setDataValue(field, value)
    }
  }
}

module.exports = {
  getSetStringify
}
