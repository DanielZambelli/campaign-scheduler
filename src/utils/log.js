const util = require('util')
const log = (object) => console.log(util.inspect(object, false, null, true))

module.exports = {log}
