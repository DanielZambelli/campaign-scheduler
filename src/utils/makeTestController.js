const { CampaignScheduler: Controller } = require('../controller/index')

const makeTestController = (id, callbacks = {}) => {

  const db = process.env.CS_DB ? JSON.parse(process.env.CS_DB) : { "dialect": "sqlite" }
  if(db.dialect==='sqlite') db.storage = `tmp/db_${id}.sqlite`

  db.schema = `test_${id}`
  db.force = true
  db.makeDestroy = true

  return new Controller({ db, callbacks }).init()
}

module.exports = {
  makeTestController
}
