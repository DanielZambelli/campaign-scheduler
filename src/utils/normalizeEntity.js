const normalizeEntity = (entity) => {
  const {createdAt,updatedAt,...entityNormalized} = entity.toJSON()
  return entityNormalized
}

module.exports = {
  normalizeEntity
}
