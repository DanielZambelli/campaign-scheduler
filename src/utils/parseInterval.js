const moment = require('moment')
const {parseIntervalOffset} = require('./parseIntervalOffset')
const {parseIntervalTime} = require('./parseIntervalTime')

const parseInterval = (interval, startDate=undefined) => {

  const {offset=null, time=null} = interval
  startDate = moment(startDate)

  //start from now or offset from previous date
  const date = moment(startDate)

  // offset
  const offsetParsed = parseIntervalOffset(offset)
  if(offsetParsed) date.add(offsetParsed)

  // specific time
  const timeParsed = parseIntervalTime(time)
  if(timeParsed) date.set(timeParsed)

  // incase now has passed the time slot offset till next day
  if(date.isSame(startDate, 'day') && startDate.isAfter(date))
    date.add({ days: 1 })

  return date.toISOString()
}

module.exports = {parseInterval}
