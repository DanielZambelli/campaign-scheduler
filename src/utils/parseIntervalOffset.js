const {getRandomBetween} = require('./getRandomBetween')

const parseIntervalOffset = (offset) => {
  if(!offset) return
  let [a,b] = offset.split('-').map(offset => {
    return offset.trim().split(' ').reduce((map,item) => {
      const value = parseInt(item.match(/\d+/g))
      const unit = item.replace(value,'')
      map[unit] = value
      return map
    }, {})
  })
  if(a && b){
    Object.keys(b).forEach(unit => {
      if(a[unit]){
        const min = Math.min(a[unit], b[unit])
        const max = Math.max(a[unit], b[unit])
        a[unit] = getRandomBetween(min,max)
      }else a[unit] = b[unit]
    })
  }
  return a
}

module.exports = {parseIntervalOffset}
