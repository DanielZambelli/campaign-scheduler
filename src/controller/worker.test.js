require('../utils/test')
const {seed} = require('../utils/test/seeds2')
const moment = require('moment')
const TEST = 'worker'

describe(TEST, () => {

  beforeEach(init(TEST,seed))

  afterEach(destroy)

  it('processes', async () => {
    await startDelayStop(1700)
    const res = await getState()
    expect(res.state.actions.filter(o => o.state === 'completed').length).toEqual(18)
    expect(invoked.length).toEqual(18)
    expect(res).toMatchSnapshot()
  })

  it('processes past dates', async () => {
    await ctl.db.Actions.update({ expectedAt: moment().subtract(2, 'hours').toDate() }, { where: {} })
    await startDelayStop(1700)
    const res = await getState()
    expect(res.state.actions.filter(o => o.state === 'completed').length).toEqual(18)
    expect(invoked.length).toEqual(18)
    expect(res).toMatchSnapshot()
  })

})
