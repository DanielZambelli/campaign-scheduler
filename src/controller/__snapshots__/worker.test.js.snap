// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`worker processes 1`] = `
{
  "hash": "cf9f9990bd79622cf895b668faa6cd2f7cbb7561",
  "state": {
    "actions": [
      {
        "actionId": "action1",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign1",
        "completedAt": "a few seconds ago",
        "expectedAt": "a few seconds ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#1",
      },
      {
        "actionId": "action2",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate2",
        },
        "campaignId": "campaign1",
        "completedAt": "a few seconds ago",
        "expectedAt": "a few seconds ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#1",
      },
      {
        "actionId": "action3",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate3",
        },
        "campaignId": "campaign1",
        "completedAt": "a few seconds ago",
        "expectedAt": "a few seconds ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#1",
      },
      {
        "actionId": "action1",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign1",
        "completedAt": "a few seconds ago",
        "expectedAt": "a few seconds ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#2",
      },
      {
        "actionId": "action2",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate2",
        },
        "campaignId": "campaign1",
        "completedAt": "a few seconds ago",
        "expectedAt": "a few seconds ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#2",
      },
      {
        "actionId": "action3",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate3",
        },
        "campaignId": "campaign1",
        "completedAt": "a few seconds ago",
        "expectedAt": "a few seconds ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#2",
      },
      {
        "actionId": "action1",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign3",
        "completedAt": "a few seconds ago",
        "expectedAt": "a few seconds ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#7",
      },
      {
        "actionId": "action2",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate2",
        },
        "campaignId": "campaign3",
        "completedAt": "a few seconds ago",
        "expectedAt": "a few seconds ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#7",
      },
      {
        "actionId": "action3",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate3",
        },
        "campaignId": "campaign3",
        "completedAt": "a few seconds ago",
        "expectedAt": "a few seconds ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#7",
      },
      {
        "actionId": "action4",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate4",
        },
        "campaignId": "campaign3",
        "completedAt": "a few seconds ago",
        "expectedAt": "a few seconds ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#7",
      },
      {
        "actionId": "action5",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate5",
        },
        "campaignId": "campaign3",
        "completedAt": "a few seconds ago",
        "expectedAt": "a few seconds ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#7",
      },
      {
        "actionId": "action6",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate6",
        },
        "campaignId": "campaign3",
        "completedAt": "a few seconds ago",
        "expectedAt": "a few seconds ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#7",
      },
      {
        "actionId": "action1",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign3",
        "completedAt": "a few seconds ago",
        "expectedAt": "a few seconds ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#8",
      },
      {
        "actionId": "action2",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate2",
        },
        "campaignId": "campaign3",
        "completedAt": "a few seconds ago",
        "expectedAt": "a few seconds ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#8",
      },
      {
        "actionId": "action3",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate3",
        },
        "campaignId": "campaign3",
        "completedAt": "a few seconds ago",
        "expectedAt": "a few seconds ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#8",
      },
      {
        "actionId": "action4",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate4",
        },
        "campaignId": "campaign3",
        "completedAt": "a few seconds ago",
        "expectedAt": "a few seconds ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#8",
      },
      {
        "actionId": "action5",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate5",
        },
        "campaignId": "campaign3",
        "completedAt": "a few seconds ago",
        "expectedAt": "a few seconds ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#8",
      },
      {
        "actionId": "action6",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate6",
        },
        "campaignId": "campaign3",
        "completedAt": "a few seconds ago",
        "expectedAt": "a few seconds ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#8",
      },
    ],
    "campaigns": [
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action1",
            "interval": {
              "offset": "0seconds",
            },
          },
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate2",
            },
            "id": "action2",
            "interval": {
              "offset": "0seconds",
            },
          },
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate3",
            },
            "id": "action3",
            "interval": {
              "offset": "0seconds",
            },
          },
        ],
        "active": true,
        "id": "campaign1",
      },
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action1",
            "interval": {
              "offset": "1hours",
            },
          },
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate2",
            },
            "id": "action2",
            "interval": {
              "offset": "2hours",
            },
          },
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate3",
            },
            "id": "action3",
            "interval": {
              "offset": "3hours",
            },
          },
        ],
        "active": false,
        "id": "campaign2",
      },
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action1",
            "interval": {
              "offset": "0seconds",
            },
          },
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate2",
            },
            "id": "action2",
            "interval": {
              "offset": "1seconds",
            },
          },
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate3",
            },
            "id": "action3",
            "interval": {
              "offset": "1.1seconds",
            },
          },
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate4",
            },
            "id": "action4",
            "interval": {
              "offset": "1.2seconds",
            },
          },
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate5",
            },
            "id": "action5",
            "interval": {
              "offset": "1.3seconds",
            },
          },
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate6",
            },
            "id": "action6",
            "interval": {
              "offset": "1.4seconds",
            },
          },
        ],
        "active": true,
        "id": "campaign3",
      },
    ],
    "schedules": [
      {
        "active": true,
        "campaignId": "campaign1",
        "subject": "contact#1",
      },
      {
        "active": true,
        "campaignId": "campaign1",
        "subject": "contact#2",
      },
      {
        "active": false,
        "campaignId": "campaign1",
        "subject": "contact#3",
      },
      {
        "active": false,
        "campaignId": "campaign1",
        "subject": "contact#4",
      },
      {
        "active": true,
        "campaignId": "campaign2",
        "subject": "contact#5",
      },
      {
        "active": true,
        "campaignId": "campaign2",
        "subject": "contact#6",
      },
      {
        "active": true,
        "campaignId": "campaign3",
        "subject": "contact#7",
      },
      {
        "active": true,
        "campaignId": "campaign3",
        "subject": "contact#8",
      },
    ],
  },
}
`;

exports[`worker processes past dates 1`] = `
{
  "hash": "6430bd4a4a60a6ed4113b8d2cbd0c7e73daa8a56",
  "state": {
    "actions": [
      {
        "actionId": "action1",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign1",
        "completedAt": "a few seconds ago",
        "expectedAt": "2 hours ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#1",
      },
      {
        "actionId": "action2",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate2",
        },
        "campaignId": "campaign1",
        "completedAt": "a few seconds ago",
        "expectedAt": "2 hours ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#1",
      },
      {
        "actionId": "action3",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate3",
        },
        "campaignId": "campaign1",
        "completedAt": "a few seconds ago",
        "expectedAt": "2 hours ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#1",
      },
      {
        "actionId": "action1",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign1",
        "completedAt": "a few seconds ago",
        "expectedAt": "2 hours ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#2",
      },
      {
        "actionId": "action2",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate2",
        },
        "campaignId": "campaign1",
        "completedAt": "a few seconds ago",
        "expectedAt": "2 hours ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#2",
      },
      {
        "actionId": "action3",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate3",
        },
        "campaignId": "campaign1",
        "completedAt": "a few seconds ago",
        "expectedAt": "2 hours ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#2",
      },
      {
        "actionId": "action1",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign3",
        "completedAt": "a few seconds ago",
        "expectedAt": "2 hours ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#7",
      },
      {
        "actionId": "action2",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate2",
        },
        "campaignId": "campaign3",
        "completedAt": "a few seconds ago",
        "expectedAt": "2 hours ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#7",
      },
      {
        "actionId": "action3",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate3",
        },
        "campaignId": "campaign3",
        "completedAt": "a few seconds ago",
        "expectedAt": "2 hours ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#7",
      },
      {
        "actionId": "action4",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate4",
        },
        "campaignId": "campaign3",
        "completedAt": "a few seconds ago",
        "expectedAt": "2 hours ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#7",
      },
      {
        "actionId": "action5",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate5",
        },
        "campaignId": "campaign3",
        "completedAt": "a few seconds ago",
        "expectedAt": "2 hours ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#7",
      },
      {
        "actionId": "action6",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate6",
        },
        "campaignId": "campaign3",
        "completedAt": "a few seconds ago",
        "expectedAt": "2 hours ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#7",
      },
      {
        "actionId": "action1",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign3",
        "completedAt": "a few seconds ago",
        "expectedAt": "2 hours ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#8",
      },
      {
        "actionId": "action2",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate2",
        },
        "campaignId": "campaign3",
        "completedAt": "a few seconds ago",
        "expectedAt": "2 hours ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#8",
      },
      {
        "actionId": "action3",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate3",
        },
        "campaignId": "campaign3",
        "completedAt": "a few seconds ago",
        "expectedAt": "2 hours ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#8",
      },
      {
        "actionId": "action4",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate4",
        },
        "campaignId": "campaign3",
        "completedAt": "a few seconds ago",
        "expectedAt": "2 hours ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#8",
      },
      {
        "actionId": "action5",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate5",
        },
        "campaignId": "campaign3",
        "completedAt": "a few seconds ago",
        "expectedAt": "2 hours ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#8",
      },
      {
        "actionId": "action6",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate6",
        },
        "campaignId": "campaign3",
        "completedAt": "a few seconds ago",
        "expectedAt": "2 hours ago",
        "log": "successfully invoked",
        "state": "completed",
        "subject": "contact#8",
      },
    ],
    "campaigns": [
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action1",
            "interval": {
              "offset": "0seconds",
            },
          },
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate2",
            },
            "id": "action2",
            "interval": {
              "offset": "0seconds",
            },
          },
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate3",
            },
            "id": "action3",
            "interval": {
              "offset": "0seconds",
            },
          },
        ],
        "active": true,
        "id": "campaign1",
      },
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action1",
            "interval": {
              "offset": "1hours",
            },
          },
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate2",
            },
            "id": "action2",
            "interval": {
              "offset": "2hours",
            },
          },
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate3",
            },
            "id": "action3",
            "interval": {
              "offset": "3hours",
            },
          },
        ],
        "active": false,
        "id": "campaign2",
      },
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action1",
            "interval": {
              "offset": "0seconds",
            },
          },
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate2",
            },
            "id": "action2",
            "interval": {
              "offset": "1seconds",
            },
          },
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate3",
            },
            "id": "action3",
            "interval": {
              "offset": "1.1seconds",
            },
          },
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate4",
            },
            "id": "action4",
            "interval": {
              "offset": "1.2seconds",
            },
          },
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate5",
            },
            "id": "action5",
            "interval": {
              "offset": "1.3seconds",
            },
          },
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate6",
            },
            "id": "action6",
            "interval": {
              "offset": "1.4seconds",
            },
          },
        ],
        "active": true,
        "id": "campaign3",
      },
    ],
    "schedules": [
      {
        "active": true,
        "campaignId": "campaign1",
        "subject": "contact#1",
      },
      {
        "active": true,
        "campaignId": "campaign1",
        "subject": "contact#2",
      },
      {
        "active": false,
        "campaignId": "campaign1",
        "subject": "contact#3",
      },
      {
        "active": false,
        "campaignId": "campaign1",
        "subject": "contact#4",
      },
      {
        "active": true,
        "campaignId": "campaign2",
        "subject": "contact#5",
      },
      {
        "active": true,
        "campaignId": "campaign2",
        "subject": "contact#6",
      },
      {
        "active": true,
        "campaignId": "campaign3",
        "subject": "contact#7",
      },
      {
        "active": true,
        "campaignId": "campaign3",
        "subject": "contact#8",
      },
    ],
  },
}
`;
